### PR TITLE
machines: Disks with custom file paths

### PR DESCRIPTION
**Pull request by skobyda**
Until this moment, a VM's disk could been only specified by **storage pool + volume combination**. However libvirt allows also to specify a disk thru a **file path** (useful for things such as adding iso files as disks to existing VM https://github.com/cockpit-project/cockpit/issues/13003).

The "Use Existing" section now has option to choose Source Type. Pool+Volume combination is default.
Screenshot from 2020-02-10 15-29-36

Custom path option allows user to specify a path to file to be used as disk.
Screenshot from 2020-02-10 15-29-18

### CODE REVIEW

**Comment by skobyda** [comment]
Ups didn't notice inconsistend case in "Source type". Change that.

File: test/verify/machineslib.py [hunk 1]
```diff
@@ -663,10 +666,16 @@ def fill(self):
                     if self.permanent:
                         b.click("#vm-{0}-disks-adddisk-permanent".format(self.vm_name))
                 else:
-                    # Choose storage pool
-                    b.select_from_dropdown("#vm-{0}-disks-adddisk-existing-select-pool".format(self.vm_name), self.pool_name)
-                    # Select from the available volumes
-                    b.select_from_dropdown("#vm-{0}-disks-adddisk-existing-select-volume".format(self.vm_name), self.volume_name)
+                    if self.source_type == "file":
```
Comment (KKoukiou) [inline test/verify/machineslib.py]: This option won't exist in images where libvirt-dbus is not present. Please add a check for that. Ex;
```
if m.image in ['ubuntu-1804]:
   wait_not_present(...
```

**Comment by garrett** [comment]
Dropdowns generally shouldn't be used when there are only <= 2 options. (It becomes arguable when there are 3 or 4.)

However, having the radios above for create new / use existing and then having radios immediately under for another radio switcher is also weird, as adding a second set may look like they're all related or that the radios are vertical options.

"Source" isn't really the source. It's really if it will be created or not. Perhaps we should rename that.

And then make "Pool" actually "Source", with a listing of pools, a divider, and also "Custom path".

BTW: How many pools are typically on a system?

**Comment by skobyda** [comment]
> Dropdowns generally shouldn't be used when there are only <= 2 options. (It becomes arguable when there are 3 or 4.)
> 
> However, having the radios above for create new / use existing and then having radios immediately under for another radio switcher is also weird, as adding a second set may look like they're all related or that the radios are vertical options.
> 
> "Source" isn't really the source. It's really if it will be created or not. Perhaps we should rename that.
> 
> And then make "Pool" actually "Source", with a listing of pools, a divider, and also "Custom path".
>
I cannot kinda imagine that. You mean that both "Pool" and "Custom path" would be visible at the same time? You don't specify pool when you have custom path.
Could you show me some mockup?

> BTW: How many pools are typically on a system?
That's depend on the user. It's like asking how many Virtual machines are typically on a system.
An user which does not do anything with Storage Pool would usually have only one storage pool: default. On the other hand, user which does a lot of storage configuration could many storage pools with many different types: iscsi, newtork filesystem, physical...
So basically anything from 1 to _a lot of_

File: test/verify/machineslib.py (hunk 1 above)
Comment (skobyda) [reply test/verify/machineslib.py]: Fixed

**Comment by garrett** [comment]
> I cannot kinda imagine that. You mean that both "Pool" and "Custom path" would be visible at the same time? You don't specify pool when you have custom path.
Could you show me some mockup?

Sure. It would take a little time to make.

However, perhaps I just wasn't being clear enough.

I meant there should be one dropdown that combines pools and the custom path for the _actual_ "Source". (The rest of the elements in the dialog would dynamically change based on what's selected in the source, much like we have now.)

It would be a [grouped select](https://www.patternfly.org/v4/documentation/react/components/formselect#grouped), like:

- Pools
  - Cockpit default
  - My first pool
  - Swimming
  - Kiddie
  - Pond
  - Lake
- _---_
- Custom path

The reason why I asked how many pools there are _typically_ is due to placing custom path at the end. If there are usually way too many, perhaps we may want to consider placing it first and setting the default to whatever was selected previously (using a cookie) and if nothing was explicitly selected, then defaulting to whatever Cockpit / libvirt uses by default for its pool.

**Comment by skobyda** [comment]
@garrett I tried to update the UI according to your suggestions. Let me know if this is what you imaged:


### The grouped selector:
Screenshot from 2020-02-14 12-18-07

### After choosing a pole, user can decide between Creating a new volume or using existing volume..
Screenshot from 2020-02-14 12-18-21
Screenshot from 2020-02-14 12-17-51

### Custom Path:
Screenshot from 2020-02-14 12-18-01

**Comment by garrett** [comment]
That looks about right. I wonder if the custom path should be in a group of 1 just to make it look a little more consistent.

What's up with the widget rendering, however? It looks very odd. The height and alignment is wrong and the caret is too large. (I'll look into this.)

The radio buttons in-between the dropdowns also seem a little strange.

Perhaps we should merge the "create new" and "use existing" in the same way, with "create new" as an option and the volumes as a group? What do you think?

**Comment by skobyda** [comment]
I don't know if combining all combinations into one big dropdown with list of all pools wouldn't be too complicated and vast if user has many storage pools or for user who does not understand fully libvirt's concepts of pools and custom paths...
Perhaps have 3 radio buttons together with options: Create new, Use existing Volume, Custom Path with a list of pools separate?

**Comment by skobyda** [comment]
A summary of discussion if virt person:

ISO file could be attached as both CDROM and disk, and we should not limit either. There are pros and cons for both.

- con: CDROM cannot be hotplugged. If you want to hotplug iso file, you can add it as a disk file or eject/load "CD" for existing cdrom device (see next point)

- pro: CDROM has a functionality to eject/load "CD" (basically change a file path for existing CDROM device), which is used and requested functionality by users why to use CDROM

- On the other hand windows guest might not like iso file as a disk file

So we should not limit iso file to either cdroms nor disks.

**Comment by garrett** [comment]
What functional difference is there be between hotplugging a disk or loading/ejecting a disc?

Why not just treat ISO files as a disc that can be loaded and ejected?

(We should not bother someone with implementation details that don't matter.)

**Comment by garrett** [comment]
Why don't we just add a "Custom path" as a source radio option, so there is "Create new" (in a pool), "Use existing" (from a pool), and "Custom path" (on the filesystem)?

And then we can simply have an input / file selector widget thingy and then under that is the "Image type" which defaults to the auto-detected best type. If there's more than one, it's a dropdown (like "CD/DVD Disc image (ISO)" and "Hard disk image (IMG)" for ISO files, but defaulting to a CD disc)? For standard disk images that aren't ISO, it would just say "Hard disk image". If it cannot be detected, then "Unknown". (And I guess you couldn't add it...?)

Does that make sense? Then we get rid of having to overload the dropdowns, get rid of having to have multiple layers of radio buttons, and so on.

So, a quick visual:

## Hard disk image

Add Disk |  
--: | --
Source | ○ Create New   ○ Use Existing   ● Custom path
Path | [ /path/to/disk.img_______________________▼]
Type | Hard disk image
  |  
  | [Cancel] [**Add**]

## Disc (CD/DVD ISO)

Add Disk |  
--: | --
Source | ○ Create New   ○ Use Existing  ● Custom path
Path | [ /path/to/disc.iso_______________________▼]
Type | [ CD/DVD Disc (ISO)  ▼]
  |  
  | [Cancel] [**Add**]

**Comment by skobyda** [comment]
> What functional difference is there be between hotplugging a disk or loading/ejecting a disc?
> 
> Why not just treat ISO files as a disc that can be loaded and ejected?
> 
> (We should not bother someone with implementation details that don't matter.)

Disks and cdroms are different on the side of guest OS (i.e. Windows guest might just straight refuse to boot from ISO which is not on cdrom, virtual disk has) + there are other differences in like cdrom is emulated and so forth.

Hotplugging can be used for disk device. Loading and ejecting is used for cdroms. They are different operations. You cannot hotplug new CDrom, but you can change media (load or eject CD) during runtime, which is.

**Comment by garrett** [comment]
> Disks and cdroms are different on the side of guest OS (i.e. Windows guest might just straight refuse to boot from ISO which is not on cdrom, virtual disk has) + there are other differences in like cdrom is emulated.

Right, but every OS can boot from an ISO as a CD-rom, so it makes sense to prioritize that.

> Hotplugging can be used for disk device. Loading and ejecting is used for cdroms. They are different operations. You cannot hotplug new CDrom, but you can change media (load or eject CD) during runtime, which is.

Again, how are people using ISO files versus normal disks? They read and boot from the ISO file. They write to disks (but not discs). So, this, with the fact that Windows won't even recognize an ISO as a hard disk means we really should support ISOs as CD/DVD discs, not disk images.

Even still, I did add a dropdown in my mockup to allow someone to change it, but I think that's probably a bad idea still. We should definitely support an ISO as a CD/DVD-ROM disc by default no matter what, however.

I still haven't seen any good reason or need for mounting an ISO as a hard disk instead of a CD/DVD. ("Because you _can_" is not a good reason.)

**Comment by skobyda** [comment]
As I said, there are pros and cons for both. It's true that user mainly wants to use cdrom for iso files, but downside with cdrom is it's inability to hotplug. So if you want to hotplug iso, and do have not attached any cdrom before running the VM, you cannot use eject/load, a so only other options is to hotplug.
Also other downside of cdrom is the fact it's emulated, and therefore slower. 

So because there are pros and cons to both, I believe we should offer option for user to choose, but change set cdrom defautly when we detect iso (as you suggested).

**Comment by skobyda** [comment]
> And then we can simply have an input / file selector widget thingy and then under that is the "Image type" which defaults to the auto-detected best type. If there's more than one, it's a dropdown (like "CD/DVD Disc image (ISO)" and "Hard disk image (IMG)" for ISO files, but defaulting to a CD disc)? For standard disk images that aren't ISO, it would just say "Hard disk image". If it cannot be detected, then "Unknown". (And I guess you couldn't add it...?)

I can only think of auto-detecting stuff based on file extensions. And file extensions in linux do not really mean anything, files could just not have them (I have a lot of discs like that). We could default to certain options according to file extensions, but not limit anything based on something as unreliable as file extensions.

**Comment by garrett** [comment]
> I can only think of auto-detecting stuff based on file extensions. And file extensions in linux do not really mean anything, files could just not have them (I have a lot of discs like that).

Well, there's also [magic](https://linux.die.net/man/5/magic). [`file`](https://linux.die.net/man/1/file) uses magic, specifically from `/usr/share/misc/magic`. Is there any way to run `file` or something like it on the file?

Falling back to extensions is not so bad. Most people use them most of the time, I suppose.

A very basic implementation would have disk first, then would switch to a CD/DVD disc if it "knows" it is an ISO, possibly by ending with an `.iso` extension. This would probably make things work in an optimal way most of the time.

Are there any other formats we'd need to care about other than disk and disc? :wink:

**Comment by skobyda** [comment]
Updated UI. I would recommend to call the dropdown a "Device" instead of "Type", since type of disk in libvirt/qemu terminology means something else and it could only confuse users. Device is proper qemu/libvirt terminology for that.

Screenshot from 2020-02-28 00-40-58

When user chooses an iso file, Device gets changed to cdrom option. Also cdrom option is incompatible with default virtio bus type, so when cdrom option is detected, bus type is changed to scsi and virtio option is disabled.
Screenshot from 2020-02-28 00-44-46

**Comment by skobyda** [comment]
> A very basic implementation would have disk first, then would switch to a CD/DVD disc if it "knows" it is an ISO, possibly by ending with an `.iso` extension. This would probably make things work in an optimal way most of the time.

Agree

> Are there any other formats we'd need to care about other than disk and disc?

Currently we support only "disk" device, this would introduce "cdrom" device. There are others like "floppy" or "lun", but nobody seems to demand them so far, so for now we can just stick to these two.

**Comment by skobyda** [comment]
@KKoukiou have you seen any of these test failures? I cannot think how they would relate to this PR, so I wonder if those are flakes.

**Comment by KKoukiou** [comment]
> @KKoukiou have you seen any of these test failures? I cannot think how they would relate to this PR, so I wonder if those are flakes.

Yes - they seem flakes. Pls retry once more anyway the tests.

**Comment by garrett** [comment]
BTW: The wonkiness of the file selector widget has already been filed @ #12409. Don't let that stop this PR though. :wink:

**Comment by skobyda** [comment]
The missing id is because we use PF3 TypeAheadSelect component, to which we pass id parameter (in this case `vm-centos-disks-adddisk-file`), but it doesn't seem to reflect anywhere. I cannot find documentation for this component whetever passing id to it is valid. 
So yeah, switch to PF4 here would fix that.

**Review by martinpitt — CHANGES_REQUESTED** [review]
Needs rebase and test fixing, so nacking for now to get it off the "needs review" list. This still looks like a great feature to get in, though!

**Comment by marusak** [comment]
File selector has been updated. @skobyda in a week this will have it's one year anniversary :tada: 
This seems very close and it is just being rebased and postponed. Do you need help with it?

File: pkg/machines/components/vm/disks/diskAdd.jsx [hunk 2]
```diff
@@ -366,10 +397,11 @@ export class AddDiskModalBody extends React.Component {
                     if (poolName)
                         stateDelta = { ...stateDelta, ...this.existingVolumeNameDelta(this.getDefaultVolumeName(poolName), prevState.storagePoolName) };
                 }
+                if (value === CUSTOM_PATH) // user moved to CUSTOM_PATH subtab
```
Comment (KKoukiou) [inline pkg/machines/components/vm/disks/diskAdd.jsx:400]: I think it's cleaner to always assign 'stadeDelte.mode = value' on the top of this case, and then do the per mode modifications.

File: pkg/machines/components/vm/disks/diskAdd.jsx [hunk 3]
```diff
@@ -379,6 +411,26 @@ export class AddDiskModalBody extends React.Component {
             this.setState({ busType: value });
             break;
         }
+        case 'file': {
+            stateDelta.file = value;
+            this.setState(prevState => {
+                if (value.endsWith(".iso"))
+                    this.onValueChanged("device", "cdrom");
```
Comment (KKoukiou) [inline pkg/machines/components/vm/disks/diskAdd.jsx:418]: Hmm this maybe would worth a comment why you don't use stateDelta here.
As in: use onValueChange instead of setState in order to perform dependent state changes

File: pkg/machines/components/vm/disks/diskAdd.jsx [hunk 4]
```diff
@@ -379,6 +411,26 @@ export class AddDiskModalBody extends React.Component {
...
+            });
+            break;
+        }
+        case 'device': {
+            stateDelta.device = value;
+            this.setState(prevState => {
+                // Disk device "cdrom" and bus "virtio" are uncompatible, see:
+                // https://www.redhat.com/archives/libvir-list/2019-January/msg01104.html
+                if (value === "cdrom" && prevState.busType === "virtio")
+                    this.onValueChanged("busType", "scsi");
```
Comment (KKoukiou) [inline pkg/machines/components/vm/disks/diskAdd.jsx:429]: virt-manager chooses SATA by default. What's the background between this deviation?

File: test/verify/check-machines-disks [hunk 5]
```diff
@@ -651,6 +668,49 @@ class TestMachinesDisks(VirtualMachinesCase):
...
+        # Prepare file for Custom File disk type
+        m.execute("touch /var/lib/libvirt/novell.iso")
+
+        self.login_and_go("/machines")
+        b.wait_in_text("body", "Virtual machines")
+
+        b.wait_in_text("#vm-subVmTest1-state", "Running")
+        self.goToVmPage("subVmTest1")
+
+        if "debian" not in m.image and "ubuntu" not in m.image:
```
Comment (KKoukiou) [inline test/verify/check-machines-disks:686]: Why debian and ubuntu can't attach cdroms?

File: test/verify/check-machines-disks [hunk 6]
```diff
@@ -651,6 +668,49 @@ class TestMachinesDisks(VirtualMachinesCase):
...
+            # Run this test only for shut off VM
```
Comment (KKoukiou) [inline test/verify/check-machines-disks:687]: This comment confuses me, you say run this test only for shut off Vms, but then you run it with running Vm, and the is not some xfail error message or anything. What i am missing?

File: test/verify/check-machines-disks [hunk 7]
```diff
@@ -651,6 +668,49 @@ class TestMachinesDisks(VirtualMachinesCase):
         if m.image in ["debian-testing"]:
             self.allow_journal_messages('.* type=1400 .* apparmor="DENIED" operation="open" profile="libvirt.* name="%s.*' % self.vm_tmpdir)
 
+    def testAddDiskCustomPath(self):
```
Comment (KKoukiou) [inline test/verify/check-machines-disks:671]: How about adding also the scenario of an non ISO file to the test.

**Review by KKoukiou — CHANGES_REQUESTED** [review]

File: pkg/machines/components/vm/disks/diskAdd.jsx (hunk 4 above)
Comment (skobyda) [reply pkg/machines/components/vm/disks/diskAdd.jsx:429]: What virtmanager does is look if a a disk with same device type exists. If so, virtmanager uses its bus (that approach I can use too). If not, it just let's the first bus form the alphabetically ordered list of compatible buses. (that I din't think I should do)

source: https://github.com/virt-manager/virt-manager/blob/master/virtManager/addhardware.py#L1056

File: test/verify/check-machines-disks (hunk 6 above)
Comment (skobyda) [reply test/verify/check-machines-disks:687]: Ahh, that comment was meant to be for other test case. This one doesn't need to have a comment. removing

File: pkg/machines/components/vm/disks/diskAdd.jsx (hunk 4 above)
Comment (skobyda) [reply pkg/machines/components/vm/disks/diskAdd.jsx:429]: Also I found this: "If omitted, the bus type is inferred from the style of the device name (e.g. a device named 'sda' will typically be exported using a SCSI bus)". It makes me feel like scsi bus should be default option since libvirt defaults to it too.
source: https://libvirt.org/formatdomain.html#hard-drives-floppy-disks-cdroms

So my idea is to combine what virtmanager does and what lib doc suggests:
- Look if a disk with same device already exists on the VM. If so, use the same bus.
- If no such disk exists, and use is switching from cdrom with bustype virtio to another device type where virtio bus type is not present, then default to scsi. (as per libvirt dov)
- else If user is switching between device types, where already selected bus types is present for both device types, then keep that bus type.

Comment (KKoukiou) [reply pkg/machines/components/vm/disks/diskAdd.jsx:429]: Sounds good. You can also just allow the user to select from the available bus types and default to the alphabetically first choice. It's up to you.
Feel free to to that as a follow up as well, how it is now is not wrong.

File: test/verify/check-machines-disks (hunk 5 above)
Comment (skobyda) [reply test/verify/check-machines-disks]: I think there was a problem year ago with apparmor not allowing custom file paths for disks. But I don't really remember it anymore. I run works well. So I deleted the condition

File: pkg/machines/components/vm/disks/diskAdd.jsx [hunk 8]
```diff
@@ -234,6 +241,31 @@ const UseExistingDisk = ({ idPrefix, onValueChanged, validationFailed, dialogVal
...
+        <FormGroup id={`${idPrefix}-device`}
+                   fieldId={`${idPrefix}-select-device`}
+                   label={_("Device")}>
+            <FormSelect id={`${idPrefix}-select-device`}
+                        onChange={value => onValueChanged('device', value)}
+                        value={dialogValues.device}>
+                <FormSelectOption value="disk" key="disk"
+                                  label={_("Hard disk image")} />
+                <FormSelectOption value="cdrom" key="cdrom"
+                                  label={_("CD/DVD disc")} />
```
Comment (KKoukiou) [inline pkg/machines/components/vm/disks/diskAdd.jsx:263]: s/disc/disk

**Review by KKoukiou — CHANGES_REQUESTED** [review]
You made a tiny mistake, other than that this is good to go in.

File: pkg/machines/components/vm/disks/diskAdd.jsx (hunk 8 above)
Comment (skobyda) [reply pkg/machines/components/vm/disks/diskAdd.jsx:263]: Actually disc is correct. It seems that both varians are correct, but "disc" is now used mostly associated with optical disc (so CD and DVDs and bluray): https://en.wikipedia.org/wiki/Spelling_of_disc
You can find the spelling used in relation to DVD: https://en.wikipedia.org/wiki/DVD
Even the word CD is a short of "compact disc": https://en.wikipedia.org/wiki/Compact_disc

Comment (KKoukiou) [reply pkg/machines/components/vm/disks/diskAdd.jsx:263]: On you got me here. However CD and DVD both have the disc in the names already, compact disc (CD) for as mention above in the link.
So adding a second disc string sounds like repeating ourselves.

I am more leaning CD/DVD device.

And the option one line above could be just 'Disk device'. What do you think?

Comment (skobyda) [reply pkg/machines/components/vm/disks/diskAdd.jsx:263]: I would shorten it to just "CD/DVD", but perhaps keep the "Hard disk" (get rid of image), since there are other types of disk devices, like floppy disk which libvirt allows too.

Comment (KKoukiou) [reply pkg/machines/components/vm/disks/diskAdd.jsx:263]: actually now that I re-read your comments I am happy with the working as is

**Review by KKoukiou — DISMISSED** [review]

**Review by KKoukiou — APPROVED** [review]

**Comment by skobyda** [comment]
New screenshots for release notes
Screenshot from 2021-02-17 12-40-47
Screenshot from 2021-02-17 12-41-41
